XL/FS: Rewrite in new format.

harshavardhana opened the pull request.

Commit:
XL/FS: Rewrite in new format.

harshavardhana committed:
No need to delete file inside erasure code (#1732)

Commit:
Listmultipart tests.

krishnasrinivas committed:
XL/erasure: fix for skipping 0 padding. (#1737)

Fixes #1736

krishnasrinivas committed:
XL/mutltipart: fix partnumber to partname association. (#1739)

Fixes #1738

harshavardhana committed:
XL: Multipart update uploads.json properly. (#1741)

harshavardhana committed:
XL: Add new metadata for checksum. (#1743)

harshavardhana committed:
XL/erasure: Reset dataBlocks to reduce the memory usage. (#1749)

Fixes #1748

harshavardhana committed:
XL: Add additional PartNumber variable as part of `xl.json` (#1750)

This is needed for verification of incoming parts and to
support variadic part uploads. Which should be sorted
properly.

Fixes #1740

harshavardhana committed:
XL/fs: initObjectLayer should cleanup tmpMetaPrefix in parallel. (#1752)

Fixes #1747

harshavardhana committed:
XL: Abortmultipart should update `uploads.json` properly. (#1757)

krishnasrinivas committed:
Fix list-incomplete uploads for XL.

abperiasamy commented on docs/backend/json-files/fs/format.json:
```diff
@@ -0,0 +1,4 @@
+{
+  "format": "fs",
+  "version": "1"
```
We should move these JSON files into a single Markdown file. I can document the backend format.

krisis committed:
Return error for empty parts in multipartupload complete (#1758)

harshavardhana committed:
XL: explicit deleteObject is not needed after rename failure. (#1760)

Reason is renameObject() does deleteObject() upon writeQuorum
failure if not keeps the successfully renamed parts if we have
reached readQuorum.

harshavardhana committed:
XL: Bring in support for object versions written during writeQuorum. (#1762)

Erasure is initialized as needed depending on the quorum and onlineDisks.
This way we can manage the quorum at the object layer.

harshavardhana committed:
XL: Rename, cleanup and add more comments. (#1769)

- xl-v1-bucket.go - removes a whole bunch of code.
- {xl-v1,fs-v1}-metadata.go - add a lot of comments and rename functions
   appropriately.

krishnasrinivas committed:
Multipart: Disable FS tests and certain test cases for list-incomplete-uploads.

harshavardhana committed:
XL: Introduce new API StorageInfo. (#1770)

This is necessary for calculating the total storage
capacity from object layer. This value is also needed for
browser UI.

Buckets used to carry this information, this patch
deprecates this feature.

krishnasrinivas committed:
XL/Multipart: isMultipartUpload() checks for presence of uploads.json on a random disk.

krishnasrinivas committed:
FS/ListMultipart: Fix FS list-multipart to work for unit test cases.

harshavardhana committed:
XL: Implement strided erasure distribution. (#1772)

Strided erasure distribution uses a new randomized
block distribution for each Put operation. This
information is captured inside `xl.json` for subsequent
Get operations.

balamurugana committed:
pkg/safe: remove temporary file on failure (#1774)

krisis committed:
Rename back multipart objects if read/write Quorum was unavailable (#1773)

krisis committed:
Fixed race during parallel PutObjectPart requests (#1775)

The race is between two parallel PutObjectPart requests updating partsInfo in xl.json.
Previously, it was being updated under a RLock().

harshavardhana committed:
erasure: read only dataBlocks if we have enough. (#1776)

Reconstruct with parity blocks if we don't have enough data blocks.

harshavardhana committed:
XL: ListObjects should not list when delimiter and prefix are '/'. (#1777)

Commit:
XL/Multipart: check existance upload uploadID after lock. (#1778)

Fixes #1767

Commit:
XL/ListParts: take the size from xl.json instead of backend file size as it will be different. (#1781)

Fixes #1779

hackintoshrao committed:
Adding defer to the lock (#1785)

krisis committed:
Remove parts that are missing in CompleteMultipartUpload (#1786)

* Remove parts that are missing in CompleteMultipartUpload

* Moved isUploadIDExists under proper namespace locks

* Moved code that deletes part files to a function

harshavardhana committed:
XL: bring in new storage API. (#1780)

Fixes #1771

harshavardhana committed:
XL: Truly use unique id's in temp directory. (#1790)

This also helps in avoiding cleaning up directories after.

Additionally this patch also fixes the problem of Range offsets.

harshavardhana committed:
XL: Fix GetObject erasure decode issues. (#1793)

Commit:
Nslock fixes (#1803)

* XL/Multipart: Support parallel upload of parts by doing NS locking appropriately.

* XL/Multipart: hold lock on the multipart upload while aborting.

krisis committed:
Handled possible short writes to httpResponseWriter (#1804)

* XL: Handled possible short writes to httpResponseWriter

* Added tests for Range Header combinations

hackintoshrao committed:
Test for ListObjectParts. (#1802)

harshavardhana committed:
XL: Cleanup and add more comments. (#1807)

harshavardhana committed:
Web: Update with ui changes. (#1808)

hackintoshrao committed:
Adding read nad write timeout for unresponsive client connectinos (#1809)

Commit:
XL/Multipart: During list-multipart-uploads ignore errFileNotFound and errDiskNotFound errors. (#1813)

Fixes #1795

Commit:
XL/Multipart: Introduce "deleted" field for uploads.json (#1810)

To future proof backend in case #1805 becomes an issue.

balamurugana committed:
erasure: allocate buffer only for non-nil disk (#1811)

harshavardhana committed:
XL: Bring in sha512 checksum support. (#1797)

Commit:
List Objects version 2. (#1815)

object: List Objects v2 support

balamurugana committed:
server: fix to have readable timeout value (#1823)

Commit:
XL/heal: heal missing format.json on replaced drives. (#1828)

fixes #1817

harshavardhana committed:
XL: Cleanup, comments and all the updated functions. (#1830)

Commit:
XL: Re-align the code again.

harshavardhana committed:
FS: Cleanup and Fix all multipart related operations. (#1836)

Commit:
FS/Multipart: Lock() to avoid race during PutObjectPart. (#1842)

Fixes #1839

Commit:
FS/Multipart: remove uploads.json on complete-multipart if no more uploadIDs are present for the object. (#1843)

Fixes #1835

harshavardhana committed:
XL: Bring in some modularity into format verification and healing. (#1832)

Commit:
XL/PutObject: Calculate size if not provided by the client and update xl.json with the correct size. (#1844)
